tests/Doctrine/Tests/ORM/Functional/HydrateAsCollectionTest.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional;
+
+use Doctrine\ORM\Tools\SchemaTool;
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Internal\Hydration\ObjectHydrator;
+use Doctrine\ORM\Internal\Hydration\SimpleObjectHydrator;
+use Doctrine\Common\Collections\ArrayCollection;
+
+use Doctrine\Tests\Models\CMS\CmsUser;
+
+class HydrateAsCollectionTest extends \Doctrine\Tests\OrmFunctionalTestCase
+{
+    /**
+     * @var CustomEntityManager
+     */
+    protected $_cem;
+
+    protected function setUp()
+    {
+        $this->useModelSet('cms');
+        parent::setUp();
+        
+        $this->_cem = CustomEntityManager::fromEm($this->_em);
+    }
+
+    public function testLoadAsCollections()
+    {
+        $user = new CmsUser;
+        $user->name = 'Guilherme';
+        $user->username = 'gblanco';
+        $user->status = 'developer';
+        $this->_cem->persist($user);
+        $this->_cem->flush();
+        
+        $userId = $user->getId();
+        
+        $this->_cem->clear();
+        
+        $repos = $this->_cem->getRepository('Doctrine\Tests\Models\CMS\CmsUser');
+        
+        $user = $repos->find($userId);
+        $this->assertInstanceOf('Doctrine\Tests\Models\CMS\CmsUser', $user);
+
+        // Test for DDC-3987
+        $user = $repos->find(9001);
+        $this->assertNull($user);
+    }
+
+}
+
+/**
+ * Simulate an array collection class from other projects that
+ * has a different implementation on the ArrayAccess methods.
+ *
+ * {@inheritDoc}
+ */
+class SomeArrayCollection extends ArrayCollection
```

Comment:
Yes, because the method offsetGet() in doctrine's ArrayCollection uses a safe guard. The reason I added "SomeArrayCollection" is to implement a custom offsetGet() method that does not check for out of bound access and causes a notice in these cases.
As I said, what the third party library does there might not be a wanted use-case for doctrine, therefore I'm not sure if this pull-request is acceptable or not. This test case just tries to mimic the setup the third party library uses including a collection class implementation from another framework.